### PR TITLE
Use ULARGE_INTEGER to get time value from FILETIME on Windows

### PR DESCRIPTION
**Pull request by thallium**
Fixes: https://github.com/eclipse-omr/omr/issues/7504

### CODE REVIEW

File: thread/common/thrprof.c [hunk 1]
```diff
@@ -121,14 +121,30 @@ omrthread_get_cpu_time_ex(omrthread_t thread, int64_t *cpuTime)
...
+		memset(&kernelTime, 0, sizeof(kernelTime));
+		memset(&userTime, 0, sizeof(userTime));
 
 		/* WARNING! Not supported on Win95!  Need to test to ensure this fails gracefully */
 
 		if (GetThreadTimes(thread->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
-			totalTime = ((int64_t)kernelTime.dwLowDateTime | ((int64_t)kernelTime.dwHighDateTime << 32))
-						+ ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
+			ULARGE_INTEGER kernelTimeInt64;
+			ULARGE_INTEGER userTimeInt64;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:139]: Suggest `U64` suffix; uses of `memset()` are missing.

File: thread/common/thrprof.c [hunk 2]
```diff
@@ -121,14 +121,30 @@ omrthread_get_cpu_time_ex(omrthread_t thread, int64_t *cpuTime)
...
+
+			kernelTimeInt64.LowPart = kernelTime.dwLowDateTime;
+			kernelTimeInt64.HighPart = kernelTime.dwHighDateTime;
+
+			userTimeInt64.LowPart = userTime.dwLowDateTime;
+			userTimeInt64.HighPart = userTime.dwHighDateTime;
+
+			totalTime = kernelTimeInt64.QuadPart + userTimeInt64.QuadPart;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:147]: Missing cast:
```c
			totalTime = (int64_t)(kernelTimeU64.QuadPart + userTimeU64.QuadPart);
```

File: thread/common/thrprof.c [hunk 3]
```diff
@@ -409,13 +425,25 @@ omrthread_get_user_time(omrthread_t thread)
...
+	memset(&creationTime, 0, sizeof(creationTime));
+	memset(&exitTime, 0, sizeof(exitTime));
+	memset(&kernelTime, 0, sizeof(kernelTime));
+	memset(&userTime, 0, sizeof(userTime));
 
 	/* WARNING! Not supported on Win95!  Need to test to ensure this fails gracefully */
 
 	if (GetThreadTimes(thread->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
-		totalTime = ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
+		ULARGE_INTEGER userTimeInt64;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:442]: Suffix, `memset()`.

File: thread/common/thrprof.c [hunk 4]
```diff
@@ -409,13 +425,25 @@ omrthread_get_user_time(omrthread_t thread)
...
+
+		userTimeInt64.LowPart = userTime.dwLowDateTime;
+		userTimeInt64.HighPart = userTime.dwHighDateTime;
+		totalTime = userTimeInt64.QuadPart;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:446]: Missing cast to `int64_t`.

File: thread/common/thrprof.c [hunk 5]
```diff
@@ -673,12 +701,28 @@ int64_t
...
+	memset(&creationTime, 0, sizeof(creationTime));
+	memset(&exitTime, 0, sizeof(exitTime));
+	memset(&kernelTime, 0, sizeof(kernelTime));
+	memset(&userTime, 0, sizeof(userTime));
 
 	if (GetProcessTimes(GetCurrentProcess(), &creationTime, &exitTime, &kernelTime, &userTime)) {
-		totalTime = ((int64_t)kernelTime.dwLowDateTime | ((int64_t)kernelTime.dwHighDateTime << 32))
-					+ ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
+		ULARGE_INTEGER kernelTimeInt64;
+		ULARGE_INTEGER userTimeInt64;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:717]: `U64` suffix, missing `memset()` calls.

File: thread/common/thrprof.c [hunk 6]
```diff
@@ -673,12 +701,28 @@ int64_t
...
+
+		kernelTimeInt64.LowPart = kernelTime.dwLowDateTime;
+		kernelTimeInt64.HighPart = kernelTime.dwHighDateTime;
+
+		userTimeInt64.LowPart = userTime.dwLowDateTime;
+		userTimeInt64.HighPart = userTime.dwHighDateTime;
+
+		totalTime = kernelTimeInt64.QuadPart + userTimeInt64.QuadPart;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:725]: Missing cast:
```c
		totalTime = (int64_t)(kernelTimeU64.QuadPart + userTimeU64.QuadPart);
```

File: thread/common/thrprof.c [hunk 7]
```diff
@@ -715,10 +759,17 @@ omrthread_get_process_times(omrthread_process_time_t *processTime)
 		 * WARNING: GetProcessTimes does not exist on pre-Win 98
 		 */
 		if (GetProcessTimes(GetCurrentProcess(), &creationTime, &exitTime, &systemTime, &userTime)) {
-			processTime->_userTime = GET_PROCESS_TIMES_IN_NANO *
-									 ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
-			processTime->_systemTime = GET_PROCESS_TIMES_IN_NANO *
-									   ((int64_t)systemTime.dwLowDateTime | ((int64_t)systemTime.dwHighDateTime << 32));
+			ULARGE_INTEGER systemTimeInt64;
+			ULARGE_INTEGER userTimeInt64;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:763]: Suffix, `memset()`.

File: thread/common/thrprof.c [hunk 8]
```diff
@@ -715,10 +759,17 @@ omrthread_get_process_times(omrthread_process_time_t *processTime)
...
+
+			systemTimeInt64.LowPart = systemTime.dwLowDateTime;
+			systemTimeInt64.HighPart = systemTime.dwHighDateTime;
+
+			userTimeInt64.LowPart = userTime.dwLowDateTime;
+			userTimeInt64.HighPart = userTime.dwHighDateTime;
+
+			processTime->_userTime = GET_PROCESS_TIMES_IN_NANO * userTimeInt64.QuadPart;
+			processTime->_systemTime = GET_PROCESS_TIMES_IN_NANO * systemTimeInt64.QuadPart;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:772]: Missing casts:
```c
			processTime->_userTime = (int64_t)(GET_PROCESS_TIMES_IN_NANO * userTimeU64.QuadPart);
			processTime->_systemTime = (int64_t)(GET_PROCESS_TIMES_IN_NANO * systemTimeU64.QuadPart);
```

File: thread/common/thrprof.c [hunk 9]
```diff
@@ -1060,9 +1111,18 @@ omrthread_get_thread_times(omrthread_thread_time_t *threadTime)
 	memset(&userTime, 0, sizeof(userTime));
 
 	if (GetThreadTimes(self->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
+		ULARGE_INTEGER kernelTimeInt64;
+		ULARGE_INTEGER userTimeInt64;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:1115]: Suffix, `memset()`.

File: thread/common/thrprof.c [hunk 10]
```diff
@@ -1060,9 +1111,18 @@ omrthread_get_thread_times(omrthread_thread_time_t *threadTime)
...
+		kernelTimeInt64.HighPart = kernelTime.dwHighDateTime;
+
+		userTimeInt64.LowPart = userTime.dwLowDateTime;
+		userTimeInt64.HighPart = userTime.dwHighDateTime;
+
 		/* Time is in 100's of nanos. Convert to nanos. */
-		threadTime->sysTime = ((int64_t)kernelTime.dwLowDateTime | ((int64_t)kernelTime.dwHighDateTime << 32)) * 100;
-		threadTime->userTime = ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32)) * 100;
+		threadTime->sysTime = kernelTimeInt64.QuadPart * 100;
+		threadTime->userTime = userTimeInt64.QuadPart * 100;
```
Comment (keithc-ca) [inline thread/common/thrprof.c:1125]: Missing casts:
```c
		threadTime->sysTime = (int64_t)(kernelTimeU64.QuadPart * 100);
		threadTime->userTime = (int64_t)(userTimeU64.QuadPart * 100);
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: port/win32/omrsysinfo.c [hunk 11]
```diff
@@ -1228,12 +1228,22 @@ omrsysinfo_get_CPU_utilization(struct OMRPortLibrary *portLibrary, struct J9Sysi
 {
```
Comment (keithc-ca) [inline port/win32/omrsysinfo.c:1229]: While you're here, could you delete this (useless) blank line?

File: thread/common/thrprof.c [hunk 12]
```diff
@@ -121,14 +121,32 @@ omrthread_get_cpu_time_ex(omrthread_t thread, int64_t *cpuTime)
...
 
 		/* WARNING! Not supported on Win95!  Need to test to ensure this fails gracefully */
 
 		if (GetThreadTimes(thread->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
-			totalTime = ((int64_t)kernelTime.dwLowDateTime | ((int64_t)kernelTime.dwHighDateTime << 32))
-						+ ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
+			ULARGE_INTEGER kernelTimeU64;
+			ULARGE_INTEGER userTimeU64;
+			memset(&kernelTime, 0, sizeof(kernelTime));
+			memset(&userTime, 0, sizeof(userTime));
```
Comment (keithc-ca) [inline thread/common/thrprof.c:141]: These initialize the wrong variables.
```c
			memset(&kernelTimeU64, 0, sizeof(kernelTimeU64));
			memset(&userTimeU64, 0, sizeof(userTimeU64));
```
Also a blank line between the declarations and these calls would be welcome (see e.g. line 129).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: thread/common/thrprof.c (hunk 12 above)
Comment (keithc-ca) [reply thread/common/thrprof.c]: > a blank line between the declarations and these calls would be welcome

Please add that blank line (for consistency with other parts of this change).

Comment (thallium) [reply thread/common/thrprof.c]: Sorry I forgot about that.

File: thread/common/thrprof.c [hunk 13]
```diff
@@ -409,13 +428,26 @@ omrthread_get_user_time(omrthread_t thread)
...
+	memset(&exitTime, 0, sizeof(exitTime));
+	memset(&kernelTime, 0, sizeof(kernelTime));
+	memset(&userTime, 0, sizeof(userTime));
 
 	/* WARNING! Not supported on Win95!  Need to test to ensure this fails gracefully */
 
 	if (GetThreadTimes(thread->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
-		totalTime = ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32));
+		ULARGE_INTEGER userTimeU64;
+		memset(&userTimeU64, 0, sizeof(userTimeU64));
```
Comment (keithc-ca) [inline thread/common/thrprof.c:446]: Please ensure there's a blank line before each group of `memset()` calls for separation from the declarations; here and before lines 722, 770, and 1124.

Comment (thallium) [reply thread/common/thrprof.c]: fixed

**Review by keithc-ca — APPROVED** [review]

**Comment by babsingh** [comment]
jenkins build all

**Review by babsingh — APPROVED** [review]

**Pull request merged**